src/nvim/api/vim.c

```diff
@@ -1245,6 +1245,88 @@ ArrayOf(Dictionary) nvim_get_keymap(String mode)
   return keymap_array(mode, NULL);
 }
 
+/// Sets a |mapping| for the given mode.
+///
+/// Note that isn't possible to specify a target buffer when map_args contains
+/// "<buffer>"; like with |:map-<buffer>|, only the current buffer will be
+/// affected.
+///
+/// @param  map_cmd   |:map-command| for this mapping, e.g. :nmap, :unmap!,
+///                   :vnoremap, etc.
+/// @param  map_args  |:map-arguments| like "<buffer>", "<nowait>", etc.
+///                   concatenated into a single string.
+/// @param  lhs   Left-hand-side |{lhs}| of the mapping.
+/// @param  rhs   Right-hand-side |{rhs}| of the mapping.
+///
+/// @returns Zero on success, nonzero on failure.
+Integer nvim_set_keymap(String map_cmd, String map_args,
+                        String lhs, String rhs)
+  FUNC_API_SINCE(6)  // TODO(Yilin-Yang): make sure this is correct
+{
+  if (!(map_cmd.size && lhs.size)) { goto RETURN_FAILURE; }
+
+  // TODO(Yilin-Yang): need much better string validation/parsing,
+  // checking for :noremap vs. :no (should be equivalent), etc.
+
+  // convert mode shortname into struct mapblock's integer representation
+  char_u *p = (char_u *)map_cmd.data;
+  int is_unmap = map_cmd.size && p[0] == 'u';
+  if (is_unmap) {  // scooch past the 'u'
+    p++;
+    map_cmd.size--;
+  } else if (rhs.size == 0) {  // must have nonempty rhs, since this isn't unmap
+    goto RETURN_FAILURE;
+  }
+  int map_excl = STRNCMP(p, "map!", 4) == 0;
+  int mode_val = get_map_mode(&p, map_excl);
+  int is_noremap = STRNCMP(p + 1, "noremap", 7) == 0;
+
+  // "unnoremap"/etc. isn't a real command
+  if (is_noremap && is_unmap) { goto RETURN_FAILURE; }
```

Comment:
No one-liners, please. :)

Reply:
That's fair, haha! No worries, I'll fix that before I RFC!